pkg/debuginfo/store.go

```diff
@@ -306,25 +286,72 @@ func (s *Store) upload(ctx context.Context, buildID, hash string, r io.Reader) e
 
 	if err := elfutils.ValidateHeader(b); err != nil {
 		// Failed to validate. Mark the incoming stream as corrupted, and let the client try to upload it again.
-		if err := s.metadataManager.markAsCorrupted(ctx, buildID); err != nil {
+		if err := s.metadata.MarkAsCorrupted(ctx, buildID); err != nil {
 			err = fmt.Errorf("failed to update metadata after uploaded, as corrupted: %w", err)
 			return status.Error(codes.Internal, err.Error())
 		}
 		return status.Error(codes.InvalidArgument, err.Error())
 	}
 
-	if err := s.metadataManager.markAsUploaded(ctx, buildID, hash); err != nil {
+	if err := s.metadata.MarkAsUploaded(ctx, buildID, hash); err != nil {
 		err = fmt.Errorf("failed to update metadata after uploaded: %w", err)
 		return status.Error(codes.Internal, err.Error())
 	}
 
 	return nil
 }
 
-func isStale(metadataFile *metadata) bool {
+func isStale(metadataFile *Metadata) bool {
 	return time.Now().Add(-15 * time.Minute).After(time.Unix(metadataFile.UploadStartedAt, 0))
 }
 
+func (s *Store) Download(req *debuginfopb.DownloadRequest, stream debuginfopb.DebugInfoService_DownloadServer) error {
```

Comment:
Should we fetch the metadata here and log an error if the state is not "uploaded" as this might indicate an issue?

Reply:
Very good point, we should even be more granular in the gRPC return codes! In regards to metadata should we then make it:

* Corrupted: `FAILED_PRECONDITION`
* Unknown, Uploaded, we just attempt to download
* Uploading: `UNAVAILABLE`

What do you think?

Reply:
This would be perfect! Just for context for others checking out this PR, we will download on 'unknown' as older Parca installations did not generate metadata